### PR TITLE
Added proposal for moving to index-header binary format.

### PR DESCRIPTION
**Pull request by bwplotka**
Some verifications in TODO, otherwise. Feedback welcome.

Definitely not a silver bullet here, but rather change in a good direction (:

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+4. causes unnecessary cache for unused blocks.
+
+This design is trying to address those 4 problems. 
+
+## Goals
+    
+* Reduce confusion between index-cache.json and IndexCache for series and postings.
+* Do not keep large pieces of the blocks (e.g symbols, label values) in the memory for all blocks. Be able to load it quickly in query time from disk.
+* Use constant amount of memory for building index-header/index-cache; download only required pieces of index from bucket and compute new TOC file.
+* Leverage mmap and let OS to unload unused pages from the physical memory as needed to satisfy the queries against the block.
```
Comment (metalmatze) [inline docs/proposals/201912_thanos_binary_index_header.md]:
```suggestion
* Leverage mmap and let OS unload unused pages from the physical memory as needed to satisfy the queries against the block.
```

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+ This is due to random access against label values and postings offsets.
+
+## No Goals
+
+* Full, no initial startup as designed in [Cortex, no initial block sync](https://github.com/thanos-io/thanos/issues/1813)
```
Comment (metalmatze) [inline docs/proposals/201912_thanos_binary_index_header.md]: I don't really get this sentence. Could we rephrase it?

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  * However this proposal might a step towards that as we might be able to load index-cache/index quickly on demand from disk.
+  * In the same time be able to load `index-header` in query time directly from bucket is not a goal of this proposal.
```
Comment (metalmatze) [inline docs/proposals/201912_thanos_binary_index_header.md]:
```suggestion
  * At the same time be able to load `index-header` in query time directly from bucket is not a goal of this proposal.
```

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+* blocks older than X will be loaded on demand on query time and kept cached until evicted.
+* Background eviction process unloads blocks without pending reads to the amount of Y (e.g 100 blocks) in LRU fashion. 
+ 
+Both X and Y being configurable.
+ 
+## Risks
+
+### Unexpected memory usage
+
+Users cares the most about surprising spikes in memory usage. Currently Store Gateway caches the whole index-cache.json. While it's silly to do so for all blocks, this
```
Comment (metalmatze) [inline docs/proposals/201912_thanos_binary_index_header.md]:
```suggestion
Users care the most about surprising spikes in memory usage. Currently, the Store Gateway caches the whole index-cache.json. While it's silly to do so for all blocks, this 
```

**Comment by brancz** [comment]
Love the gist of this. I haven't had the chance to do a very thorough review, but I think this is only natural to do. Nice job!

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+Step 1: Replace `index-cache.json` with `index-header`: 
+  * Implement building `index-header` file from pieces of TSDB index.
+  * Allow Store Gateway to build it instead of `index-cache.json` to disk.
+  * Allow Store Gateway to use `index-header` instead of `index-cache.json` for queries.
+  * Load those files using mmap
+  * Remove building `index-cache.json` in compactor.
+Step 2: Load on demand in query time:
+  * Allow Store Gateway to load `index-header` on demand from disk in query time.
+  * Add background job to unmmap/unload `index-header` for blocks that are unused (LRU).
```
Comment (GiedriusS) [inline docs/proposals/201912_thanos_binary_index_header.md]: Not just unused but the least recently used blocks if the memory limit is violated, right?

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+The process for building this will be as follows:
+
+* Thanks to https://github.com/thanos-io/thanos/pull/1792 we can check final size of index and scan for TOC file.
+* With TOC:
+  * Get the symbols table and copy it to the local file.
+  * Get the posting table and copy it to the local file.
+  * Get the label offset table and copy it to the local file.
+  * Get the label index and copy it to the local file.
+  
+With that effort building time and resources should be compared with downloading the prebuild `index-header` from bucket. This allows us to reduce complexity of the system as
```
Comment (GiedriusS) [inline docs/proposals/201912_thanos_binary_index_header.md]:
```suggestion
With that effort building time and resources should be compared with downloading the prebuilt `index-header` from the bucket. This allows us to reduce complexity of the system as 
```

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+How to micro-benchmark such change? mmaping is outside of Go run time, which counts allocations etc.
+ 
+### Do we need to mmap?
+
+mmap adds a lot of complexity and confusion especially around monitoring it's memory usage as it does not appear on Go profiles.
+
+While mmap is great for random access against a big file, in fact the current implementation of the [FileReader](https://github.com/prometheus/prometheus/blob/de0a772b8e7d27dc744810a1a693d97be027049a/tsdb/index/index.go#L664)
+reallocates symbols, offsets and label name=value pairs while reading. This really defies the purpose of mmap as we want to combine all info in a dense few sequential sections of index-header binary format, 
+this file will be mostly read sequentially. Still, label values can be accessed randomly, that's why we propose to start with mmap straight away.
```
Comment (GiedriusS) [inline docs/proposals/201912_thanos_binary_index_header.md]: But with this, I guess, we would be tied to platforms which support mmap(2). That would exclude Windows AFAIK. It's OK with me, just something to be aware of.

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+* https://github.com/thanos-io/thanos/issues/1711 (pulling index in smarter way)
+* https://github.com/thanos-io/thanos/pull/1013 (initial investigation of different formats)
+* https://github.com/thanos-io/thanos/issues/448 (OOM)
+* https://github.com/thanos-io/thanos/issues/1705 (LTS umbrella issue)
+
+## Summary
+
+This short document describes the motivation and the design of a new format that is meant to replace `index-cache.json` we have currently.
+
+We also propose renaming index-cache to *index-header* due to name collision with index cache for postings and series.
```
Comment (GiedriusS) [inline docs/proposals/201912_thanos_binary_index_header.md]: 👍

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+## Verification
+
+* Compare memory and latency usage for startup without index-header being generated on compactor
+* Benchmark for loading the `index-header` vs `index-cache.json` into Store GW.
+
+## Proposal
+
+TSDB index is in binary [format](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md). 
+
+To allow reduced resource consumption and effort when building (1), (2), "index-header" for blocks we plan to reuse similar format for sections like symbols, label indices and
```
Comment (GiedriusS) [inline docs/proposals/201912_thanos_binary_index_header.md]: Not sure if I have missed this detail but what is the actual format that this proposes? Protobufs?

File: docs/proposals/201912_thanos_binary_index_header.md (hunk 7 above)
Comment (brancz) [reply docs/proposals/201912_thanos_binary_index_header.md]: In prometheus this seems to work: https://github.com/prometheus/prometheus/blob/master/tsdb/fileutil/mmap_windows.go

Comment (GiedriusS) [reply docs/proposals/201912_thanos_binary_index_header.md]: TIL about that, that's nice! We should definitely re-use that :p disregard this message, then.

Comment (bwplotka) [reply docs/proposals/201912_thanos_binary_index_header.md]: All is good on this front.

File: docs/proposals/201912_thanos_binary_index_header.md (hunk 5 above)
Comment (bwplotka) [reply docs/proposals/201912_thanos_binary_index_header.md]: How we would know this?

File: docs/proposals/201912_thanos_binary_index_header.md (hunk 9 above)
Comment (bwplotka) [reply docs/proposals/201912_thanos_binary_index_header.md]: A custom binary format was planned, composed with those sections as mentioned, however, I might find a blocker to this ):

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+* blocks newer then X (e.g 2w) being always loaded
+* blocks older than X will be loaded on demand on query time and kept cached until evicted.
+* background eviction process unloads blocks without pending reads to the amount of Y (e.g 100 blocks) in LRU fashion. 
+ 
+Both X and Y being configurable.
+ 
+## Risks
+
+### Posting size to fetch
```
Comment (bwplotka) [inline docs/proposals/201912_thanos_binary_index_header.md]: Added this which we missed initially in this design. Let me know if it's clear. 

TL;DR: we scan `postings` form TSDB index when building `index-cache` and without changing TSDB index this will be still required building `index-header`

@brancz @GiedriusS

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+This means maintaining pool with limited number of `index-header` files loaded at the time in Store Gateway.
+With an LRU logic we should be able to deduce which of blocks should be unloaded and left on the disk. 
+
+This proposes following algorithm:
+
+* blocks newer then X (e.g 2w) being always loaded
+* blocks older than X will be loaded on demand on query time and kept cached until evicted.
+* background eviction process unloads blocks without pending reads to the amount of Y (e.g 100 blocks) in LRU fashion. 
+ 
+Both X and Y being configurable.
```
Comment (pracucci) [inline docs/proposals/201912_thanos_binary_index_header.md]: I understand the proposed algorithm would be easier to implement, but the only real downside of keeping all `index-header` loaded is the memory limit. From the UX perspective would be great being able to set a memory limit for the `index-header` and having Thanos automatically adjusting X and Y based on actual indexes size.

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+makes this proposal less valuable as we still need to download big part of index and traverse through it instead of what we propose in [#Proposal](./201912_thanos_binary_index_header.md#proposal)
+
+For series we don't know the exact size either, however we estimate max size of each series to be 64*1024. It depends on sane number of label pairs and chunks per series.
+We have really only one potential case when this was too low: https://github.com/thanos-io/thanos/issues/552. Decision about series this was made here: https://github.com/thanos-io/thanos/issues/146
+
+For postings it's more tricky as it depends on number of series in which given label pair exists. For worst case it can be even million label-pair for popular pairs like `__name__=http_request_http_duration_bucket` etc.
+
+We have few options:
+
+* Encode this posting size in TSDB index `PostingOffset`.
```
Comment (pracucci) [inline docs/proposals/201912_thanos_binary_index_header.md]: I'm not sure this is a real option. Even if such change will be accepted and merged (and I don't think), you would have to deal with all previously created blocks, which invalidates this option.

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+
+For series we don't know the exact size either, however we estimate max size of each series to be 64*1024. It depends on sane number of label pairs and chunks per series.
+We have really only one potential case when this was too low: https://github.com/thanos-io/thanos/issues/552. Decision about series this was made here: https://github.com/thanos-io/thanos/issues/146
+
+For postings it's more tricky as it depends on number of series in which given label pair exists. For worst case it can be even million label-pair for popular pairs like `__name__=http_request_http_duration_bucket` etc.
+
+We have few options:
+
+* Encode this posting size in TSDB index `PostingOffset`. 
+* Scan postings to fetch, which is something we wanted to avoid when building `index-header` without downloading full TSDB index.
```
Comment (pracucci) [inline docs/proposals/201912_thanos_binary_index_header.md]: I genuinely believe this option just invalidates the proposal. We shouldn't download the postings section at all while building the `index-header`.

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+* Estimate some large value (will overfetch)
+* Estimate casual value and retry fetching remaining size on demand for data consistency.
```
Comment (pracucci) [inline docs/proposals/201912_thanos_binary_index_header.md]: If I'm not missing anything, the postings `End` is only used in `bucketIndexReader.fetchPostings()`, affecting only the postings which have not been fetched from the cache.

Let's consider to use a casual value. The partitioner will partition into 1+ parts, which will get fetched. Upon fetching and processing all of them, we're able to calculate the end of each posting and - if there's anyone for which some partial data is missing - we can fetch all the missing data in a second iteration, so the overhead introduced in the worst case scenario shouldn't be greater than 2x compared to today.

Given we're already over fetching due to how the max gap in the partitioner works, and given when accessing the bucket store the request latency impacts more than the bandwidth (ie. if you benchmark reading 1KB and 1MB, reading 1MB is not 1024x slower than reading 1KB) and given the work I'm doing to add memcached support for the index cache, this option doesn't look particularly bad to me, given the benefits this proposal bring.

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+
+mmap adds a lot of complexity and confusion especially around monitoring it's memory usage as it does not appear on Go profiles.
+
+While mmap is great for random access against a big file, in fact the current implementation of the [FileReader](https://github.com/prometheus/prometheus/blob/de0a772b8e7d27dc744810a1a693d97be027049a/tsdb/index/index.go#L664)
+reallocates symbols, offsets and label name=value pairs while reading. This really defies the purpose of mmap as we want to combine all info in a dense few sequential sections of index-header binary format, 
+this file will be mostly read sequentially. Still, label values can be accessed randomly, that's why we propose to start with mmap straight away. 
+
+### Is initial long startup, really a problem?
+
+After initial startup with persistent disk, next startup should be quick due to cached files on disk for old blocks.
```
Comment (pracucci) [inline docs/proposals/201912_thanos_binary_index_header.md]: > After initial startup with persistent disk, next startup should be quick due to cached files on disk for old blocks.

Another use case is scaling. When you scale up your cluster because of high load, you want the store to be ready as quickly as possible. The initial long startup delays it.

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+1, 2 & 3 contributes to Store Gateway startup being slow and resource consuming: https://github.com/thanos-io/thanos/issues/448
+
+4. causes unnecessary cache for unused blocks.
+
+This design is trying to address those 4 problems. 
+
+## Goals
+    
+* Reduce confusion between index-cache.json and IndexCache for series and postings.
+* Do not keep large pieces of the blocks (e.g symbols, label values) in the memory for all blocks. Be able to load it quickly in query time from disk.
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply docs/proposals/201912_thanos_binary_index_header.md]: will do thanks

File: docs/proposals/201912_thanos_binary_index_header.md (hunk 12 above)
Comment (bwplotka) [reply docs/proposals/201912_thanos_binary_index_header.md]: No really. We need to pull PostingOffset anyway. (:

File: docs/proposals/201912_thanos_binary_index_header.md (hunk 13 above)
Comment (bwplotka) [reply docs/proposals/201912_thanos_binary_index_header.md]: Yes fully agree

File: docs/proposals/201912_thanos_binary_index_header.md (hunk 12 above)
Comment (pracucci) [reply docs/proposals/201912_thanos_binary_index_header.md]: I thought the option was to change the TSDB index to add the size/end to the postings offset table, but I guess I'm wrong. Can you elaborate on this option, please?

File: docs/proposals/201912_thanos_binary_index_header.md (hunk 14 above)
Comment (bwplotka) [reply docs/proposals/201912_thanos_binary_index_header.md]: Thanks for this - I have similar thoughts. This is still doable indeed. I think  I agree we should take is a preffered option so far.

File: docs/proposals/201912_thanos_binary_index_header.md (hunk 15 above)
Comment (bwplotka) [reply docs/proposals/201912_thanos_binary_index_header.md]: yup, adding

File: docs/proposals/201912_thanos_binary_index_header.md (hunk 5 above)
Comment (GiedriusS) [reply docs/proposals/201912_thanos_binary_index_header.md]: Sorry, I meant some kind of limit in terms of their size on disk because it would be really nice to avoid RAM usage blowing up in case of some queries. Having all of it in RAM is horrible, yes, but OTOH it gives the users a relatively certain thing: we know that at least that part is more or less equal over a period of time.

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+* blocks older than X will be loaded on demand on query time and kept cached until evicted.
+* background eviction process unloads blocks without pending reads to the amount of Y (e.g 100 blocks) in LRU fashion. 
+ 
+Both X and Y being configurable.
+ 
+## Risks
+
+### Posting size to fetch
+
+While idea of combing different pieces of TSDB index as our index-header is great, unfortunately we heavily rely
```
Comment (pracucci) [inline docs/proposals/201912_thanos_binary_index_header.md]: Can't we infer the `End` looking at the `Start` of the next posting, sorting postings by `Start`?

For each posting we have label name, label value and offset to its series list in the postings section. The postings section contains all postings, one after the other one. If we sort the postings table by offset, the end range for the posting at position X should be equal to the start range of the posting at position X+1. The last posting will end at the position of the label index table.

Am I missing anything?

File: docs/proposals/201912_thanos_binary_index_header.md (hunk 10 above)
Comment (bwplotka) [reply docs/proposals/201912_thanos_binary_index_header.md]: We got some answers (:

**Comment by bwplotka** [comment]
Starting implementation (: 

cc @GiedriusS @pracucci if you want to take another look.

File: docs/proposals/201912_thanos_binary_index_header.md (hunk 17 above)
Comment (bwplotka) [reply docs/proposals/201912_thanos_binary_index_header.md]: As discussed offline added to the doc (:

File: docs/proposals/201912_thanos_binary_index_header.md (hunk 12 above)
Comment (bwplotka) [reply docs/proposals/201912_thanos_binary_index_header.md]: Added all of options including those not choosen

File: docs/proposals/201912_thanos_binary_index_header.md [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+
+Thanos will build/compose all index-headers on the startup for now, however in theory we can load and build those blocks on demand.
+Given the minimal memory that each loaded block should take now, this is described as [Future Work](./201912_thanos_binary_index_header.md#Future-Work)
+
+### Length of Posting to fetch
+
+While idea of combing different pieces of TSDB index as our index-header is great, unfortunately we heavily rely 
+on information about size of each posting represented as `postingRange.End`. 
+
+We need to know that apriori to know how to partition and how many bytes we need to fetch from the storage to get each posting: https://github.com/thanos-io/thanos/blob/7e11afe64af0c096743a3de8a594616abf52be45/pkg/store/bucket.go#L1567
```
Comment (pracucci) [inline docs/proposals/201912_thanos_binary_index_header.md]: > We need to know that apriori to know how to partition

I guess you mean `We need to know apriori how to partition`

Comment (bwplotka) [reply docs/proposals/201912_thanos_binary_index_header.md]: Will fix in next PR (:
